### PR TITLE
Apply declare_strict_types on codebase

### PR DESCRIPTION
**Pull request by devrck**
Fixing issue https://github.com/meilisearch/meilisearch-php/issues/59

### CODE REVIEW

**Comment by devrck** [comment]
@curquiza I find it strange why on my dev environment, following the steps of running tests it never is successful.

```bash
$ composer run-script test
> sh scripts/tests.sh
Setting FDs limit to 1000
Launching tests...
PHPUnit 8.5.8 by Sebastian Bergmann and contributors.

................................................................. 65 / 97 ( 67%)
..............F.................                                  97 / 97 (100%)

Time: 9.71 seconds, Memory: 8.00 MB

There was 1 failure:

1) Tests\Settings\DistinctAttributeTest::testUpdateDistinctAttribute
Failed asserting that 1 matches expected 'description'.

/home/devrck/projects/meilisearch-php/tests/Settings/DistinctAttributeTest.php:33

FAILURES!
Tests: 97, Assertions: 376, Failures: 1.
Script sh scripts/tests.sh handling the test event returned with error code 1

```

**Comment by curquiza** [comment]
@devrck I tried to run the test your branch, it works 🙂

Did you do run:
```bash
$ docker pull getmeili/meilisearch:latest
```
?

Because it looks like the `latest` on machine is not the `latest` on Docker Hub 😉

**Comment by devrck** [comment]
Yes that did the trick :smile: .

Btw why not wrap these commands into an composer run-script or makefile directive?

**Comment by curquiza** [comment]
> Btw why not wrap these commands into an composer run-script or makefile directive?

Hum I'm not sure I want to apply transparently definitive changes on the Docker of our contributors... I would rather add an explained line in the contributing guideline so that they exactly know was is happening 🙂

File: tests/Settings/AcceptNewFieldsTest.php [hunk 1]
```diff
@@ -8,9 +10,9 @@ class AcceptNewFieldsTest extends TestCase
 {
     private $index;
 
-    public function __construct()
+    public function setUp(): void
```
Comment (Guikingone) [inline tests/Settings/AcceptNewFieldsTest.php:13]: This method is actually `protected` in the parent class 🙂

File: tests/Settings/AttributesForFacetingTest.php [hunk 2]
```diff
@@ -1,17 +1,13 @@
 <?php
 
+declare(strict_types=1);
+
 namespace Tests\Settings;
 
 use Tests\TestCase;
 
 class AttributesForFacetingTest extends TestCase
```
Comment (Guikingone) [inline tests/Settings/AttributesForFacetingTest.php:9]: Can be final?

File: tests/Settings/DistinctAttributeTest.php [hunk 3]
```diff
@@ -8,9 +10,9 @@ class DistinctAttributeTest extends TestCase
 {
     private $index;
 
-    public function __construct()
+    public function setUp(): void
```
Comment (Guikingone) [inline tests/Settings/DistinctAttributeTest.php:13]: Same as before about visibility

File: tests/Settings/RankingRulesTest.php [hunk 4]
```diff
@@ -17,9 +19,9 @@ class RankingRulesTest extends TestCase
         'exactness',
     ];
 
-    public function __construct()
+    public function setUp(): void
```
Comment (Guikingone) [inline tests/Settings/RankingRulesTest.php:22]: Same

File: tests/Settings/StopWordsTest.php [hunk 5]
```diff
@@ -8,10 +10,9 @@ class StopWordsTest extends TestCase
 {
     private $index;
 
-    protected function setUp(): void
+    public function setUp(): void
```
Comment (Guikingone) [inline tests/Settings/StopWordsTest.php:13]: Same

File: tests/TestCase.php [hunk 6]
```diff
@@ -21,12 +23,19 @@ abstract class TestCase extends BaseTestCase
...
 
-    public function __construct()
+    public function setUp(): void
     {
-        parent::__construct();
+        parent::setUp();
         $this->client = new Client(self::HOST, self::DEFAULT_KEY);
+    }
+
+    public function tearDown(): void
```
Comment (Guikingone) [inline tests/TestCase.php:37]: That's also a `protected` method

File: src/Exceptions/InvalidArgumentException.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MeiliSearch\Exceptions;
+
+use Exception;
+
+class InvalidArgumentException extends Exception
```
Comment (Guikingone) [inline src/Exceptions/InvalidArgumentException.php:9]: Can be final

File: tests/Endpoints/ClientTest.php [hunk 8]
```diff
@@ -10,12 +12,6 @@
 
 class ClientTest extends TestCase
 {
-    protected function setUp(): void
-    {
-        parent::setUp();
-        $this->client->deleteAllIndexes();
-    }
-
     public function testGetAllIndexesWhenEmpty()
```
Comment (Guikingone) [inline tests/Endpoints/ClientTest.php:15]: Maybe we can add return typehint for each test method even if it's `: void`?

File: tests/Endpoints/DocumentsTest.php [hunk 9]
```diff
@@ -1,8 +1,11 @@
...
 
+declare(strict_types=1);
+
 namespace Tests\Endpoints;
 
 use MeiliSearch\Exceptions\HTTPRequestException;
+use MeiliSearch\Exceptions\InvalidArgumentException;
 use Tests\TestCase;
 
 class DocumentsTest extends TestCase
```
Comment (Guikingone) [inline tests/Endpoints/DocumentsTest.php:11]: Can be final

File: tests/Endpoints/ClientTest.php (hunk 8 above)
Comment (devrck) [reply tests/Endpoints/ClientTest.php:15]: This seems useless, I've added the `:void` to setUp because that's the method signature.

**Comment by curquiza** [comment]
@devrck, sorry again, but could you [rebase your branch](https://gist.github.com/curquiza/5f7ce615f85331f083cd467fc4e19398), please? 🙂

I will merge just after that, since you seem having apply avery comment of @Guikingone.

File: src/Exceptions/InvalidArgumentException.php [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+declare(strict_types=1);
+
+namespace MeiliSearch\Exceptions;
+
+use Exception;
+
+final class InvalidArgumentException extends Exception
+{
+    public static function invalidType(string $argumentName, array $validTypes)
```
Comment (Guikingone) [inline src/Exceptions/InvalidArgumentException.php:11]: Rerurn typehint

File: src/Exceptions/InvalidArgumentException.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        return new self(
+            sprintf('Argument "%s" is not a valid type! Please provide an argument that is of type: "%s"', $argumentName, \implode('","', $validTypes)),
+            400,
+            null
+        );
+    }
+
+    public static function emptyArgument(string $argumentName)
```
Comment (Guikingone) [inline src/Exceptions/InvalidArgumentException.php:20]: Return typehint

**Comment by Guikingone** [comment]
Hi 👋 

I've submitted 3 reviews for return typehint, should be good after that 🙂

**Comment by devrck** [comment]
@Guikingone could you tell me what is the value we add by adding `:void` on test cases?

**Comment by Guikingone** [comment]
@devrck The idea is to separate the tests of the data providers (which returns arrays or generator), it's easier to read and maintain if we know that every tests are executed without being dependent on other ones (which can occur if you don't specify the return typehint, we can be tempted to use this test as something else than just a test 🙂

**Comment by devrck** [comment]
@Guikingone that make sense, thank you for the explanation, it didn't cross my mind :smile: .

File: tests/TestCase.php [hunk 12]
```diff
@@ -21,16 +23,23 @@ abstract class TestCase extends BaseTestCase
 
     const DEFAULT_KEY = 'masterKey';
```
Comment (Guikingone) [inline tests/TestCase.php:24]: Consider adding a visibility for this const 🙂

Comment (devrck) [reply tests/TestCase.php]: Nice, added.

**Comment by curquiza** [comment]
Thanks @devrck for your work and your patience! And thanks also @Guikingone for your review! Really helpful!

**Pull request merged**